### PR TITLE
mimir.rules.kubernetes: Add support for clustered mode

### PR DESCRIPTION
**Pull request by 56quarters**
#### PR Description

This change adds support for updating Mimir rules based on PrometheusRule objects when running in clustered mode. It does this by picking a single instance to be the leader responsible for making changes using the Mimir API. Other non-leader instances will not process events or make calls to the Mimir API.

#### Which issue(s) this PR fixes

Related: https://github.com/grafana/alloy/pull/158
Related: https://github.com/grafana/alloy/pull/616
Related: https://github.com/grafana/alloy/pull/648

#### Notes to the Reviewer

It might be helpful to review this by commit:

- a05ac6b55e63f49ac85e29b6ebc72adfc696208e Introduces leader election such that only a single Alloy instance will receive Kubernetes events and make requests to the Mimir API based on them. This was a relatively small change but lacked any unit tests.
- 2d46f69f2800502eaccacb96553e98abe3589197 Was a larger refactor that kept the same idea as the previous commit but moved things around and introduced a few internal interfaces so that the interesting logic in `Component::Run()` could be unit tested.
- 0023f5f789bfb75343e0f78caf81efa6d5155bd7 Removes the warning in the docs introduced by #648

#### PR Checklist

- [x] CHANGELOG.md updated
- [x] Documentation added
- [x] Tests updated
- [ ] Config converters updated

### CODE REVIEW

File: internal/component/mimir/rules/kubernetes/rules.go
Comment (rfratto) [inline internal/component/mimir/rules/kubernetes/rules.go:299]: When we talked about this offline, we discussed that there will eventually be a transition to per-component scheduling where the user would opt-in components to this type of behavior. Moving to per-component scheduling could cause bugs here; notably we would want to ignore these updates if per-component scheduling was configured, since the ownership calculation of per-component scheduling may not work the same as the ownership detection here, and if both are in play there may be no Alloy instances running the component. 

Can we add a comment here notifying us of this future issue?

**Comment by 56quarters** [comment]
Question for the maintainers:

I find myself needing to do a fair amount of refactoring of this component in order to test the interesting bits of the clustering logic added to `Component::Run()`. Before I get too far along, what sort of testing would you like to see for this change? I think the component will be in a better place after this refactoring but it's also a bigger change to review.

File: internal/component/mimir/rules/kubernetes/rules.go
Comment (rfratto) [inline internal/component/mimir/rules/kubernetes/rules.go:206]: We don't make any guarantees that Update never gets called before Run, so I'm worried about the possibility of a future deadlock here if the controller gets updated. 

Since we set the component health on an error anyway, maybe we don't need to send an error back to the Update method?

File: internal/component/mimir/rules/kubernetes/rules.go
Comment (56quarters) [reply internal/component/mimir/rules/kubernetes/rules.go:215]: I'm not sure, I was preserving the existing behavior of retrying indefinitely the first time the component is started. I can change it if you'd like.

Comment (rfratto) [reply internal/component/mimir/rules/kubernetes/rules.go:215]: If that's the existing behavior I think it's fine to keep 👍

File: internal/component/mimir/rules/kubernetes/rules.go
Comment (56quarters) [reply internal/component/mimir/rules/kubernetes/rules.go:206]: Sure, I can change that. This was preserving the existing behavior of the component.

File: internal/component/mimir/rules/kubernetes/rules.go
Comment (56quarters) [reply internal/component/mimir/rules/kubernetes/rules.go:194]: > Only break if `errors.Is(err, errShutdown)`

I can do this. Would should the behavior for a non-shutdown error be, panic since returning a non-shutdown error is a bug?

File: internal/component/mimir/rules/kubernetes/rules.go
Comment (rfratto) [reply internal/component/mimir/rules/kubernetes/rules.go:206]: Thanks for pointing that out. I think it's OK to leave the other issue that was preserving behavior, but since this is a smaller change and I'm worried about it suddenly breaking I'd like to fix at least this one up.

File: internal/component/mimir/rules/kubernetes/rules.go
Comment (rfratto) [reply internal/component/mimir/rules/kubernetes/rules.go:194]: > I can do this. Would should the behavior for a non-shutdown error be, panic since returning a non-shutdown error is a bug?

I think it can be logged or ignored, potentially combined with setting health. I generally don't want components to have panics because of the blast radius; other components can still continue working even if there's a bug in one of them.

File: internal/component/mimir/rules/kubernetes/rules.go
Comment (56quarters) [reply internal/component/mimir/rules/kubernetes/rules.go]: Fixed in de9737a21629cbc064f594ef0808125d06dfdb52

File: internal/component/mimir/rules/kubernetes/rules.go
Comment (56quarters) [reply internal/component/mimir/rules/kubernetes/rules.go]: Fixed in de9737a21629cbc064f594ef0808125d06dfdb52

**Review by rfratto — APPROVED** [review]
LGTM, thanks for all the work you've put into this! 

@clayton-cornell can you check the new docs changes before we merge?

File: internal/component/mimir/rules/kubernetes/rules.go
Comment (56quarters) [reply internal/component/mimir/rules/kubernetes/rules.go]: Fixed in de9737a21629cbc064f594ef0808125d06dfdb52

File: docs/sources/reference/components/mimir.rules.kubernetes.md
Comment (clayton-cornell) [inline docs/sources/reference/components/mimir.rules.kubernetes.md:28]:
```suggestion
{{< param "PRODUCT_NAME" >}} version 1.1 and higher supports [clustered mode][] in this component. When you use this component as part
```

**Review by clayton-cornell — APPROVED** [review]
Minor rephrase and good to go

**Pull request merged**
